fish-rust/src/builtins/abbr.rs

```diff
@@ -0,0 +1,595 @@
+            "%ls %ls: Abbreviation '%ls' cannot have spaces in the word\n",
+            CMD,
+            subcmd,
+            name.as_utfstr()
+        ));
+        return STATUS_INVALID_ARGS;
+    }
+
+    let mut regex = None;
+
+    let key = if let Some(ref regex_pattern) = opts.regex_pattern {
+        // Compile the regex as given; if that succeeds then wrap it in our ^$ so it matches the
+        // entire token.
+        let flags = ffi::re::flags_t { icase: false };
+        let result = ffi::try_compile(regex_pattern, &flags);
+
+        if result.has_error() {
+            let error = result.get_error();
+            streams.err.append(wgettext_fmt!(
+                "%ls: Regular expression compile error: %ls\n",
+                CMD,
+                &error.message().from_ffi()
+            ));
+            streams
+                .err
+                .append(wgettext_fmt!("%ls: %ls\n", CMD, regex_pattern.as_utfstr()));
+            streams
+                .err
+                .append(wgettext_fmt!("%ls: %*ls\n", CMD, error.offset, "^"));
+            return STATUS_INVALID_ARGS;
+        }
+        let anchored = regex_make_anchored(regex_pattern);
+        let mut result = ffi::try_compile(&anchored, &flags);
+        assert!(
+            !result.has_error(),
+            "Anchored compilation should have succeeded"
+        );
+        let re = result.as_mut().get_regex();
+        assert!(!re.is_null(), "Anchored compilation should have succeeded");
+
+        let _ = regex.insert(re);
+        regex_pattern
+    } else {
+        // The name plays double-duty as the token to replace.
+        name
+    };
+
+    if opts.function.is_some() && opts.args.len() > 1 {
+        streams
+            .err
+            .append(wgettext_fmt!(BUILTIN_ERR_TOO_MANY_ARGUMENTS, L!("abbr")));
+        return STATUS_INVALID_ARGS;
+    }
+    let replacement = if let Some(ref function) = opts.function {
+        // Abbreviation function names disallow spaces.
+        // This is to prevent accidental usage of e.g. `--function 'string replace'`
+        if !valid_func_name(function) || contains_whitespace(function) {
+            streams.err.append(wgettext_fmt!(
+                "%ls: Invalid function name: %ls\n",
+                CMD,
+                function.as_utfstr()
+            ));
+            return STATUS_INVALID_ARGS;
+        }
+        function.clone()
+    } else {
+        let mut replacement = WString::new();
+        for iter in opts.args.iter().skip(1) {
+            if !replacement.is_empty() {
+                replacement.push(' ')
+            };
+            replacement.push_utfstr(iter);
+        }
+        replacement
+    };
+
+    let position = opts.position.unwrap_or(Position::Command);
+
+    // Note historically we have allowed overwriting existing abbreviations.
+    abbrs::with_abbrs_mut(move |abbrs| {
+        abbrs.add(Abbreviation {
+            name: name.clone(),
+            key: key.clone(),
```

Comment:
nit: we should take Options by value/move, then we could get rid of these clones.
But there's no harm in following the C++ code

Reply:
Both are already references, as key is same as name when no regex given, So clone will just move above instead of here